[move-book] Add section on unit tests in Move (#9888)
diem/diem: language/documentation/book/src/unit-testing.md

# Unit Tests

Unit testing for Move adds three new annotations to the Move source language:

* `#[test]`
* `#[test_only]`, and
* `#[expected_failure]`.

They respectively mark a function as a test, mark a module or module member (`use`, function, or struct) as code to be included for testing only, and mark that a test is expected to fail/abort. These annotations can be placed on a function with any visibility. Whenever any module or module member is annotated as `#[test_only]` or `#[test]`, it will not be included in the compiled bytecode unless it is compiled for testing.

## Testing Annotations: Their Meaning and Usage

Both the `#[test]` and `#[expected_failure]` annotations can be used either with or without arguments.

Without arguments, the `#[test]` annotation can only be placed on a function with no parameters. This annotation simply marks this function as a test to be run by the unit testing harness.

```
#[test] // OK
fun this_is_a_test() { ... }

#[test] // Will fail to compile since the test takes an argument
fun this_is_not_correct(arg: signer) { ... }
```

> A test can also be annotated as an `#[expected_failure]`. This annotation marks that the annotated test should abort — however the exact abort code is not checked. The test will be marked as failing only if the test fails with a non-abort error, or does not abort. Only functions that have the `#[test]` annotation can also be annotated as an  #`[expected_failure]`.

Reviewer: Could you provide an example, either here or in a subsequent paragraph or section, about what a "non-abort error" means? Is a bytecode verification or VM error considered a non-abort error?

If my understanding is correct, then I think I'd find this clearer as:

```suggestion
A test can also be annotated as an `#[expected_failure]`. This annotation marks that the annotated test should abort — however the exact abort code is not checked. The test will pass only if an `abort` statement is executed that halts the test's execution. Only functions that have the `#[test]` annotation can also be annotated as an  #`[expected_failure]`.
```

Author: Ah, so this also covers other errors such as if you do a `borrow_global` on a resource that doesn't exist. I'll update with an example

```
#[test]
#[expected_failure]
public fun this_test_should_abort_and_pass() { abort 1 }

#[test, expected_failure] // Can have multiple in one attribute. This test will pass.
public(script) fun this_other_test_should_abort_and_pass() { abort 1 }
```

> With arguments, a test annotation takes the form `#[test(<param_name_1> = <address>, ..., <param_name_n> = <address>)]`. If a function is annotated in such a manner, the function's parameters must be a permutation of the parameters <`param_name_1>, ..., <param_name_n>`, i.e., the order of these parameters as they occur in the function and their order in the test annotation do  not have to be the same, but they must be able to be matched up with each other by name.

Reviewer: Slight typo, looks like an extra space:
```suggestion
With arguments, a test annotation takes the form `#[test(<param_name_1> = <address>, ..., <param_name_n> = <address>)]`. If a function is annotated in such a manner, the function's parameters must be a permutation of the parameters <`param_name_1>, ..., <param_name_n>`, i.e., the order of these parameters as they occur in the function and their order in the test annotation do not have to be the same, but they must be able to be matched up with each other by name.
```

module ABC { ... }

#[test_only] // test only attributes can be attached to named addresses
address ADDR = @0x1;

#[test_only] // .. to uses
use 0x1::SomeOtherModule;

#[test_only] // .. to structs
struct SomeStruct { ... }

#[test_only] // .. and functions. Can only be called from test code, but not a test
fun test_only_function(...) { ... }
```

## Running Unit Tests

Unit tests for a Move package can be run with the [`move package test`
command](./packages.md).

When running tests, every test will either `PASS`, `FAIL`, or `TIMEOUT`. If a test case fails, the location of the failure along with the function name that caused the failure will be reported if possible. You can see an example of this below.

A test will be marked as timing out if it exceeds the maximum number of instructions that can be executed for any single test. This bound can be changed using the options below, and its default value is set to 5000 instructions. Additionally, while the result of a test is always deterministic, tests are run in parallel by default, so the ordering of test results in a test run is non-deterministic unless running with only one thread (see `OPTIONS` below).

> There are also a number of options that can be passed to the unit testing binary to fine-tune testing, and to help debug failing tests. These are:
> 
> ```
> SAGE:
>     move package test [FLAGS] [OPTIONS]
> 
> FLAGS:
>         --stackless         Use the stackless bytecode interpreter to run the tests and cross check its results with the
>                             execution result from Move VM
>         --coverage
>     -g, --state_on_error    Show the storage state at the end of execution of a failing test
>     -h, --help              Prints help information
>     -l, --list              List all tests
>     -s, --statistics        Report test statistics at the end of testing
>     -V, --version           Prints version information
>     -v                      Print additional diagnostics
>         --verbose           Verbose mode
> 
> OPTIONS:
>     -f, --filter <filter>                A filter string to determine which unit tests to run
>     -i, --instructions <instructions>    Bound the number of instructions that can be executed by any one test [default:
>                                          5000]
>     -t, --threads <num_threads>          Number of threads to use for running tests [default: 8]
>     -p, --path <package-path>            Path to a package which the command should be run with respect to. Defaults to
>                                          current directory [default: .]
>         --storage-dir <storage-dir>      Directory storing Move resources, events, and module bytecodes produced by
>                                          module publishing and script execution [default: storage]
> ```
> 
> While each of these flags and options are fairly self-explanatory, it is worth mentioning that when filtering tests with the `-f` option, any test whose fully qualified name (i.e., `address::module::function_name`) contains the `<filter>` string will be run. Additionally, the `--coverage` flag collects test coverage information that you can view using the `move package coverage` command.

Reviewer: There are a few documents in our repository that include the output from tools' `--help` options, and I've found many of them to be out of date (it's also hard to keep them in sync -- do I send pull requests to update them? That means I need to get reviews, wait for bors to land the code... and anyone who changes the options or their help text needs to remember to update any docs that includes help text). Instead, how about the following?

```suggestion
There are also a number of options that can be passed to the unit testing binary to fine-tune testing, and to help debug failing tests. You may view these by running `move package test --help`.
```

In this case, though, I would suggest that the additional notes you include here ought to be included in the documentation of the options. That is, the `-f` option help text should include the fact that any test that includes the filter string will be run, and the `--coverage` option help text should indicate that its output can be viewed with the `move package coverage` command -- if they don't already, that is.

Author: Very good point! I'll update this, and add the additional documentation to the commands

Author: I've removed this, and updated the documentation comments around the package commands in the bottom commit of this PR.
